File: src/main/java/com/chidozie/core/renderEngine/KeyInputManager.java
package com.chidozie.core.renderEngine;

import java.util.HashMap;
import java.util.Map;

public class KeyInputManager {
    private final Map<Integer, Boolean> keyStates = new HashMap<>();
    private final WindowManager windowManager;
    private final Map<Integer, Long> lastKeyPressTimes = new HashMap<>();
    private final long debounceInterval = 400; // Set debounce interval (in milliseconds)

    public KeyInputManager(WindowManager windowManager) {
        this.windowManager = windowManager;
    }

    public boolean isKeyJustPressed(int key) {
        long currentTime = System.currentTimeMillis();
        boolean isPressed = windowManager.isKeyPressed(key);
        boolean wasPressed = keyStates.getOrDefault(key, false);

        if (isPressed && !wasPressed) {
            long lastPressTime = lastKeyPressTimes.getOrDefault(key, 0L);
            if (currentTime - lastPressTime >= debounceInterval) {
                keyStates.put(key, true);
                lastKeyPressTimes.put(key, currentTime);
                return true;
            }
        } else if (!isPressed) {
            keyStates.put(key, false);
        }

        return false;
    }
}

File: src/main/kotlin/com/adv/core/renderEngine/KeyInputManager.kt
package com.adv.core.renderEngine

class KeyInputManager(private val windowManager: WindowManager) {
    private val keyStates: MutableMap<Int?, Boolean?> = HashMap<Int?, Boolean?>()
    private val lastKeyPressTimes: MutableMap<Int?, Long?> = HashMap<Int?, Long?>()
    private val debounceInterval: Long = 400 // Set debounce interval (in milliseconds)

    fun isKeyJustPressed(key: Int): Boolean {
        val currentTime = System.currentTimeMillis()
        val isPressed = windowManager.isKeyPressed(key)
        val wasPressed = keyStates.getOrDefault(key, false)!!

        if (isPressed && !wasPressed) {
            val lastPressTime = lastKeyPressTimes.getOrDefault(key, 0L)!!
            if (currentTime - lastPressTime >= debounceInterval) {
                keyStates.put(key, true)
                lastKeyPressTimes.put(key, currentTime)
                return true
            }
        } else if (!isPressed) {
            keyStates.put(key, false)
        }

        return false
    }
}
